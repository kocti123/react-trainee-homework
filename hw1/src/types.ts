import { AnyAction } from "redux";
import { ThunkAction } from "redux-thunk";

export interface ITask {
  readonly id: number;
  readonly body: string;
  readonly completed: boolean;
  readonly favourite: boolean;
  readonly isEdit: boolean;
  readonly newBody: string;
  readonly creationDate: Date;
}

export type filters = "all" | "completed" | "favourite" | "notCompleted";

export type TasksState = {
  readonly isLoaded: boolean;
  readonly lastId: number;
  readonly tasks: ITask[];
  readonly currentFilter: filters;
  readonly filtredTasks: ITask[];
};

export type State = {
  readonly tasks: TasksState;
};

export type AppAction = ThunkAction<void, State, unknown, AnyAction>;
