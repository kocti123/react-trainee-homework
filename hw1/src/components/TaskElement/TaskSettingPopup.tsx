import { useEffect, useRef } from "react";
import ReactDOM from "react-dom";
import { useAppDispatch } from "../../redux";
import {
  removeFromCompleted,
  markTaskAsCompleted,
  removeFromFavourite,
  markTaskAsFavourite,
  setIsEdit,
} from "../../redux/actions/tasksActions";
import { ITask } from "../../types";
import styles from "./TaskSettingPopup.module.css";

const popupDOM = document.querySelector<HTMLDListElement>("#popup")!;

function TaskSettingPopup({
  onClose,
  onDelete,
  task,
  id,
  position,
}: {
  onClose: () => void;
  onDelete: (task: ITask) => void;
  task: ITask;
  id: number;
  position: { x: number; y: number };
}) {
  const dispatch = useAppDispatch();
  const { completed, favourite, isEdit } = task;
  const popupRef = useRef<HTMLUListElement>(null);

  useEffect(() => {
    if (popupRef.current) {
      popupRef.current.style.top = `${position.y}px`;
      popupRef.current.style.left = `${position.x}px`;
    }
  }, [position, popupRef]);

  function toggleCompleted() {
    if (completed) {
      dispatch(removeFromCompleted(id));
    } else {
      dispatch(markTaskAsCompleted(id));
    }
    onClose();
  }

  function toggleFavourite() {
    if (favourite) {
      dispatch(removeFromFavourite(id));
    } else {
      dispatch(markTaskAsFavourite(id));
    }
    onClose();
  }

  function toggleIsEdit() {
    if (isEdit) {
      dispatch(setIsEdit(id, false));
    } else {
      dispatch(setIsEdit(id, true));
    }
    onClose();
  }

  function removeTaskHandler() {
    onDelete(task);
    onClose();
  }

  return ReactDOM.createPortal(
    <>
      <div className={styles.popupWrapper} onClick={onClose}></div>
      <ul ref={popupRef} className={styles.popup}>
        <li>
          <button onClick={toggleFavourite}>
            {favourite ? "Убрать из избранного" : "В избранное"}
          </button>
        </li>
        <li>
          <button onClick={toggleCompleted}>
            {completed ? "Вернуть в работу" : "Выполненно"}
          </button>
        </li>
        <li>
          <button onClick={toggleIsEdit}>
            {isEdit ? "Отменить редактирование" : "Редактировать"}
          </button>
        </li>
        <li>
          <button onClick={removeTaskHandler}>Удалить</button>
        </li>
      </ul>
    </>,
    popupDOM
  );
}

export default TaskSettingPopup;
