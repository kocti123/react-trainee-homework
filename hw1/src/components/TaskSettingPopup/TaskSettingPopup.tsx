import { useEffect, useRef } from "react";
import ReactDOM from "react-dom";
import { useAppDispatch } from "../../redux";
import {
  removeFromCompleted,
  markTaskAsCompleted,
  removeFromFavourite,
  markTaskAsFavourite,
  setIsEdit,
} from "../../redux/actions/tasksActions";
import { ITask } from "../../types";
import styles from "./TaskSettingPopup.module.css";

function TaskSettingPopup({
  onClose,
  onDelete,
  task,
  position,
}: {
  onClose: () => void;
  onDelete: (task: ITask) => void;
  task: ITask;
  position: { x: number; y: number };
}) {
  const dispatch = useAppDispatch();
  const { completed, favourite, isEdit, id } = task;
  const popupDOM = document.querySelector<HTMLDListElement>("#popup")!;
  const popupRef = useRef<HTMLUListElement>(null);

  useEffect(() => {
    if (popupRef.current) {
      popupRef.current.style.top = `${position.y}px`;
      popupRef.current.style.left = `${position.x}px`;
    }
  }, [position, popupRef]);

  function toggleCompleted() {
    if (completed) {
      dispatch(removeFromCompleted(id));
    } else {
      dispatch(markTaskAsCompleted(id));
    }
    onClose();
  }

  function toggleFavourite() {
    if (favourite) {
      dispatch(removeFromFavourite(id));
    } else {
      dispatch(markTaskAsFavourite(id));
    }
    onClose();
  }

  function toggleIsEdit() {
    if (isEdit) {
      dispatch(setIsEdit(id, false));
    } else {
      dispatch(setIsEdit(id, true));
    }
    onClose();
  }

  function removeTaskHandler() {
    onDelete(task);
    onClose();
  }

  return ReactDOM.createPortal(
    <>
      <div className={styles.popupWrapper} onClick={onClose}></div>
      <ul ref={popupRef} className={styles.popup}>
        <li>
          <button onClick={toggleFavourite} className="toggleFavourite">
            {favourite ? "Убрать из избранного" : "В избранное"}
          </button>
        </li>
        <li>
          <button onClick={toggleCompleted} className="toggleCompletion">
            {completed ? "Вернуть в работу" : "Выполненно"}
          </button>
        </li>
        <li>
          <button onClick={toggleIsEdit} className="toggleIsEdit">
            {isEdit ? "Отменить редактирование" : "Редактировать"}
          </button>
        </li>
        <li>
          <button onClick={removeTaskHandler} className="remove">
            Удалить
          </button>
        </li>
      </ul>
    </>,
    popupDOM
  );
}

export default TaskSettingPopup;
