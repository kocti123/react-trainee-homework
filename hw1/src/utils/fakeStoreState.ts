import { State } from "../types";

export function getFakeStoreState(): State {
  return {
    tasks: {
      isLoaded: true,
      lastId: 1,
      currentFilter: "all",
      tasks: [
        {
          id: 1,
          body: "test",
          completed: false,
          favourite: false,
          isEdit: false,
          newBody: "",
          creationDate: new Date(1635696113017),
        },
      ],
      filtredTasks: [
        {
          id: 1,
          body: "test",
          completed: false,
          favourite: false,
          isEdit: false,
          newBody: "",
          creationDate: new Date(1635696113017),
        },
      ],
    },
  };
}
