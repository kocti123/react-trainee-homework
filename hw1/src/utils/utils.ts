import { AnyAction } from "redux";
import { tasksActionTypes } from "../constants/actionTypes";
import { filters, ITask } from "../types";

export const filterTasks = (tasks: ITask[], filter: filters): ITask[] => {
  switch (filter) {
    case "all":
      return tasks;
    case "favourite":
      return filterFavouriteAndNotCompleted(tasks);
    case "completed":
      return filterCompleted(tasks);
    case "notCompleted":
      return filterNotCompleted(tasks);
    default:
      throw new Error("wrong filter type");
  }
};

export const filterFavouriteAndNotCompleted = (tasks: ITask[]): ITask[] => {
  return tasks.filter((task) => task.favourite && !task.completed);
};

export const filterCompleted = (tasks: ITask[]): ITask[] => {
  return tasks.filter((task) => task.completed);
};

export const filterNotCompleted = (tasks: ITask[]): ITask[] => {
  return tasks.filter((task) => !task.completed);
};

export const filterOutById = (tasks: ITask[], id: number): ITask[] => {
  return tasks.filter((task) => task.id !== id);
};

export const findById = (tasks: ITask[], id: number): ITask => {
  return tasks.filter((task) => task.id === id)[0];
};

export const lastIdOfTasks = (tasks: ITask[]): number => {
  return Math.max(...tasks.map((task) => task.id));
};

export const filterAndSortTasks = (
  tasks: ITask[],
  filter: filters
): ITask[] => {
  const filteredTasks = filterTasks(tasks, filter);
  return filteredTasks.sort((a, b) => b.id - a.id);
};

export const getUpdateAction = (id: number, task: ITask): AnyAction => {
  return {
    type: tasksActionTypes.UPDATE_TASK,
    payload: {
      id,
      task: {
        ...task,
      },
    },
  };
};
