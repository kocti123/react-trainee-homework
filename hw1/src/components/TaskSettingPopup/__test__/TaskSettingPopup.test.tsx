import { ReactWrapper, mount } from "enzyme";
import { Provider } from "react-redux";
import createMockStore, {
  MockStoreCreator,
  MockStoreEnhanced,
} from "redux-mock-store";
import thunk from "redux-thunk";
import { AppAction, ITask, State } from "../../../types";
import { getFakeStoreState } from "../../../utils/fakeStoreState";
import TaskSettingPopup from "../TaskSettingPopup";
import styles from "../TaskSettingPopup.module.css";

const mockActionTypes = {
  markFav: "markFav",
  markComp: "markComp",
  removeFav: "removeFav",
  removeComp: "removeComp",
  isEdit: "isEdit",
};

jest.mock("../../../redux/actions/tasksActions", () => ({
  removeFromCompleted:
    (id: number): AppAction =>
      (dispatch) => {
        dispatch({ type: mockActionTypes.removeComp, payload: id });
      },
  markTaskAsCompleted:
    (id: number): AppAction =>
      (dispatch) => {
        dispatch({ type: mockActionTypes.markComp, payload: id });
      },
  removeFromFavourite:
    (id: number): AppAction =>
      (dispatch) => {
        dispatch({ type: mockActionTypes.removeFav, payload: id });
      },
  markTaskAsFavourite:
    (id: number): AppAction =>
      (dispatch) => {
        dispatch({ type: mockActionTypes.markFav, payload: id });
      },
  setIsEdit:
    (id: number, value: boolean): AppAction =>
      (dispatch) => {
        dispatch({ type: mockActionTypes.isEdit, payload: { id, value } });
      },
}));

const createStore: MockStoreCreator<State, any> = createMockStore([thunk]);
const storeState = getFakeStoreState();

describe("Task Setting Popup", () => {
  let popupDiv: HTMLDivElement;
  let wrapper: ReactWrapper;
  let store: MockStoreEnhanced<unknown, {}>;
  const props: {
    onClose: any;
    onDelete: any;
    task: ITask;
    position: { x: number; y: number };
  } = {
    onClose: jest.fn(),
    onDelete: jest.fn((task: ITask) => { }),
    task: storeState.tasks.filtredTasks[0],
    position: {
      x: 10,
      y: 10,
    },
  };

  beforeAll(() => {
    popupDiv = document.createElement("div");
    popupDiv.id = "popup";
    document.body.append(popupDiv);
  });

  afterAll(() => {
    popupDiv.remove();
    jest.clearAllMocks();
  });

  beforeEach(() => {
    store = createStore();
    jest.spyOn(store, "dispatch");
    wrapper = mount(<TaskSettingPopup {...props} />, {
      wrappingComponent: Provider,
      wrappingComponentProps: { store: store },
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
    wrapper.unmount();
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should set inline styles based on provided position", () => {
    const popupDOM = wrapper
      .find(`.${styles.popup}`)
      .getDOMNode<HTMLUListElement>();

    expect(popupDOM.style.getPropertyValue("top")).toBe(
      `${props.position.x}px`
    );
    expect(popupDOM.style.getPropertyValue("left")).toBe(
      `${props.position.y}px`
    );
  });

  it("should change inline styles when position changed", () => {
    const popupDOM = wrapper
      .find(`.${styles.popup}`)
      .getDOMNode<HTMLUListElement>();

    wrapper.setProps({ position: { x: 20, y: 20 } });

    expect(popupDOM.style.getPropertyValue("top")).toBe("20px");
    expect(popupDOM.style.getPropertyValue("left")).toBe("20px");
  });

  it("should properly toggle favourite status", () => {
    const toggleButton = wrapper.find(".toggleFavourite");

    toggleButton.simulate("click");

    wrapper.setProps({ task: { ...props.task, favourite: true } });

    toggleButton.simulate("click");

    expect(store.dispatch).toBeCalledTimes(2);
    expect(store.getActions()).toEqual([
      { type: mockActionTypes.markFav, payload: 1 },
      { type: mockActionTypes.removeFav, payload: 1 },
    ]);
  });

  it("should have proper text in button based on favourite status", () => {
    const toggleButton = wrapper.find(".toggleFavourite");

    expect(toggleButton.text()).toBe("В избранное")

    wrapper.setProps({ task: { ...props.task, favourite: true } });

    expect(toggleButton.text()).toBe("Убрать из избранного")
  });

  it("should properly toggle completed status", () => {
    const toggleButton = wrapper.find(".toggleCompletion");

    toggleButton.simulate("click");

    wrapper.setProps({ task: { ...props.task, completed: true } });

    toggleButton.simulate("click");

    expect(store.dispatch).toBeCalledTimes(2);
    expect(store.getActions()).toEqual([
      { type: mockActionTypes.markComp, payload: 1 },
      { type: mockActionTypes.removeComp, payload: 1 },
    ]);
  });

  it("should have proper text in button based on completed status", () => {
    const toggleButton = wrapper.find(".toggleCompletion");

    expect(toggleButton.text()).toBe("Выполненно")

    wrapper.setProps({ task: { ...props.task, completed: true } });

    expect(toggleButton.text()).toBe("Вернуть в работу")
  });

  it("should properly toggle isEdit status", () => {
    const toggleButton = wrapper.find(".toggleIsEdit");

    toggleButton.simulate("click");

    wrapper.setProps({ task: { ...props.task, isEdit: true } });

    toggleButton.simulate("click");

    expect(store.dispatch).toBeCalledTimes(2);
    expect(store.getActions()).toEqual([
      { type: mockActionTypes.isEdit, payload: { id: 1, value: true } },
      { type: mockActionTypes.isEdit, payload: { id: 1, value: false } },
    ]);
  });

  it("should have proper text in button based on isEdit status", () => {
    const toggleButton = wrapper.find(".toggleIsEdit");

    expect(toggleButton.text()).toBe("Редактировать")

    wrapper.setProps({ task: { ...props.task, isEdit: true } });

    expect(toggleButton.text()).toBe("Отменить редактирование")
  });

  it("should call onDelete when click delete button", () => {
    const removeButton = wrapper.find(".remove");

    removeButton.simulate("click");

    expect(props.onDelete).toBeCalledTimes(1);
    expect(props.onDelete).toBeCalledWith(props.task);
  });

  it("should call onClose when click on any button", () => {
    wrapper.find(".toggleFavourite").simulate("click");
    wrapper.find(".toggleCompletion").simulate("click");
    wrapper.find(".toggleIsEdit").simulate("click");
    wrapper.find(".remove").simulate("click");

    expect(props.onClose).toBeCalledTimes(4);
  });

  it("should call onClose when click outside of popup", () => {
    wrapper.find(`.${styles.popupWrapper}`).simulate("click");

    expect(props.onClose).toBeCalledTimes(1);
  })
});
