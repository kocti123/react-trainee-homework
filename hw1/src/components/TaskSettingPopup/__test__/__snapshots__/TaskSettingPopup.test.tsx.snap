// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Setting Popup should match snapshot 1`] = `ReactWrapper {}`;
